File: chat/chat_manager.py
"""
Менеджер чатов для единообразной работы с разными чатами.
Реализует принципы SOLID и DRY.
"""

import logging
from typing import Dict, List, Optional, Any
from dataclasses import dataclass
from abc import ABC, abstractmethod

logger = logging.getLogger(__name__)


@dataclass
class ChatSettings:
    """Настройки чата."""
    model: str
    temperature: float
    top_p: float
    max_tokens: int
    system_prompt: str
    web_search: bool
    functions: List[Dict[str, Any]]


@dataclass
class ChatMessage:
    """Сообщение в чате."""
    id: str
    role: str
    content: str
    timestamp: str
    token_stats: Optional[Dict[str, Any]] = None


@dataclass
class ChatSession:
    """Сессия чата."""
    session_id: str
    settings: ChatSettings
    messages: List[ChatMessage]
    stats: Optional[Dict[str, Any]] = None


class ChatAPI(ABC):
    """Абстрактный класс для работы с API чатов."""
    
    @abstractmethod
    async def create_session(self, settings: ChatSettings) -> str:
        """Создает новую сессию чата."""
        pass
    
    @abstractmethod
    async def send_message(self, session_id: str, message: str, functions: List[Dict[str, Any]]) -> Dict[str, Any]:
        """Отправляет сообщение в чат."""
        pass
    
    @abstractmethod
    async def update_session_settings(self, session_id: str, settings: ChatSettings) -> bool:
        """Обновляет настройки сессии."""
        pass


class DjangoChatAPI(ChatAPI):
    """Реализация API для Django backend."""
    
    def __init__(self, csrf_token: str):
        self.csrf_token = csrf_token
        self.base_url = '/playground/api'
    
    async def create_session(self, settings: ChatSettings) -> str:
        """Создает новую сессию чата."""
        import requests
        
        response = requests.post(
            f'{self.base_url}/create-session/',
            headers={
                'Content-Type': 'application/json',
                'X-CSRFToken': self.csrf_token
            },
            json={
                'model': settings.model,
                'temperature': settings.temperature,
                'top_p': settings.top_p,
                'max_tokens': settings.max_tokens,
                'system_prompt': settings.system_prompt,
                'web_search': settings.web_search,
                'functions': settings.functions
            }
        )
        
        if response.status_code == 200:
            data = response.json()
            if data.get('success'):
                return data['session_id']
        
        raise Exception(f"Failed to create session: {response.text}")
    
    async def send_message(self, session_id: str, message: str, functions: List[Dict[str, Any]]) -> Dict[str, Any]:
        """Отправляет сообщение в чат."""
        import requests
        
        response = requests.post(
            f'{self.base_url}/send-message/',
            headers={
                'Content-Type': 'application/json',
                'X-CSRFToken': self.csrf_token
            },
            json={
                'session_id': session_id,
                'message': message,
                'functions': functions
            }
        )
        
        if response.status_code == 200:
            data = response.json()
            if data.get('success'):
                return data
        
        raise Exception(f"Failed to send message: {response.text}")
    
    async def update_session_settings(self, session_id: str, settings: ChatSettings) -> bool:
        """Обновляет настройки сессии."""
        import requests
        
        response = requests.post(
            f'{self.base_url}/update-session/',
            headers={
                'Content-Type': 'application/json',
                'X-CSRFToken': self.csrf_token
            },
            json={
                'session_id': session_id,
                'model': settings.model,
                'temperature': settings.temperature,
                'top_p': settings.top_p,
                'max_tokens': settings.max_tokens,
                'system_prompt': settings.system_prompt,
                'web_search': settings.web_search
            }
        )
        
        if response.status_code == 200:
            data = response.json()
            return data.get('success', False)
        
        return False


class ChatManager:
    """Менеджер для управления чатами."""
    
    def __init__(self, api: ChatAPI):
        self.api = api
        self.sessions: Dict[str, ChatSession] = {}
    
    def create_chat_session(self, chat_id: str, settings: ChatSettings) -> ChatSession:
        """Создает новую сессию чата."""
        session = ChatSession(
            session_id="",
            settings=settings,
            messages=[]
        )
        self.sessions[chat_id] = session
        return session
    
    async def initialize_session(self, chat_id: str) -> str:
        """Инициализирует сессию чата."""
        if chat_id not in self.sessions:
            raise Exception(f"Chat session {chat_id} not found")
        
        session = self.sessions[chat_id]
        session_id = await self.api.create_session(session.settings)
        session.session_id = session_id
        
        logger.info(f"Session initialized for chat {chat_id}: {session_id}")
        return session_id
    
    async def send_message(self, chat_id: str, message: str) -> ChatMessage:
        """Отправляет сообщение в чат."""
        if chat_id not in self.sessions:
            raise Exception(f"Chat session {chat_id} not found")
        
        session = self.sessions[chat_id]
        
        # Если нет session_id, инициализируем сессию
        if not session.session_id:
            await self.initialize_session(chat_id)
        
        # Добавляем сообщение пользователя
        user_message = ChatMessage(
            id=str(len(session.messages) + 1),
            role='user',
            content=message,
            timestamp=self._get_timestamp()
        )
        session.messages.append(user_message)
        
        # Отправляем сообщение через API
        response = await self.api.send_message(
            session.session_id,
            message,
            session.settings.functions
        )
        
        # Добавляем ответ ассистента
        assistant_message = ChatMessage(
            id=str(len(session.messages) + 1),
            role='assistant',
            content=response['assistant_message']['content'],
            timestamp=response['assistant_message']['timestamp'],
            token_stats=response['assistant_message']['token_stats']
        )
        session.messages.append(assistant_message)
        
        # Обновляем статистику
        if 'session_stats' in response:
            session.stats = response['session_stats']
        
        logger.info(f"Message sent to chat {chat_id}, response length: {len(assistant_message.content)}")
        return assistant_message
    
    async def update_settings(self, chat_id: str, settings: ChatSettings) -> bool:
        """Обновляет настройки чата."""
        if chat_id not in self.sessions:
            raise Exception(f"Chat session {chat_id} not found")
        
        session = self.sessions[chat_id]
        session.settings = settings
        
        # Если есть session_id, обновляем настройки на сервере
        if session.session_id:
            success = await self.api.update_session_settings(session.session_id, settings)
            if success:
                logger.info(f"Settings updated for chat {chat_id}")
            return success
        
        return True
    
    def get_session(self, chat_id: str) -> Optional[ChatSession]:
        """Получает сессию чата."""
        return self.sessions.get(chat_id)
    
    def clear_messages(self, chat_id: str):
        """Очищает сообщения чата."""
        if chat_id in self.sessions:
            self.sessions[chat_id].messages = []
            logger.info(f"Messages cleared for chat {chat_id}")
    
    def _get_timestamp(self) -> str:
        """Получает текущий timestamp."""
        from datetime import datetime
        return datetime.now().isoformat()


class ChatFactory:
    """Фабрика для создания чатов."""
    
    @staticmethod
    def create_chat_manager(csrf_token: str) -> ChatManager:
        """Создает менеджер чатов."""
        api = DjangoChatAPI(csrf_token)
        return ChatManager(api)
    
    @staticmethod
    def create_settings(
        model: str,
        temperature: float,
        top_p: float,
        max_tokens: int,
        system_prompt: str = "",
        web_search: bool = False,
        functions: List[Dict[str, Any]] = None
    ) -> ChatSettings:
        """Создает настройки чата."""
        if functions is None:
            functions = []
        
        return ChatSettings(
            model=model,
            temperature=temperature,
            top_p=top_p,
            max_tokens=max_tokens,
            system_prompt=system_prompt,
            web_search=web_search,
            functions=functions
        )

File: static/js/chat-manager.js
/**
 * Менеджер чатов для единообразной работы с разными чатами.
 * Реализует принципы SOLID и DRY.
 */

class ChatSettings {
    constructor(model, temperature, top_p, max_tokens, system_prompt = "", web_search = false, functions = []) {
        this.model = model;
        this.temperature = temperature;
        this.top_p = top_p;
        this.max_tokens = max_tokens;
        this.system_prompt = system_prompt;
        this.web_search = web_search;
        this.functions = functions;
    }
}

class ChatMessage {
    constructor(id, role, content, timestamp, token_stats = null) {
        this.id = id;
        this.role = role;
        this.content = content;
        this.timestamp = timestamp;
        this.token_stats = token_stats;
    }
}

class ChatSession {
    constructor(session_id, settings, messages = [], stats = null) {
        this.session_id = session_id;
        this.settings = settings;
        this.messages = messages;
        this.stats = stats;
    }
}

class ChatAPI {
    constructor(csrf_token) {
        this.csrf_token = csrf_token;
        this.base_url = '/playground/api';
    }

    async createSession(settings) {
        const response = await fetch(`${this.base_url}/create-session/`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'X-CSRFToken': this.csrf_token
            },
            body: JSON.stringify({
                model: settings.model,
                temperature: settings.temperature,
                top_p: settings.top_p,
                max_tokens: settings.max_tokens,
                system_prompt: settings.system_prompt,
                web_search: settings.web_search,
                functions: settings.functions
            })
        });

        if (response.ok) {
            const data = await response.json();
            if (data.success) {
                return data.session_id;
            }
        }

        throw new Error(`Failed to create session: ${response.statusText}`);
    }

    async sendMessage(session_id, message, functions) {
        const response = await fetch(`${this.base_url}/send-message/`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'X-CSRFToken': this.csrf_token
            },
            body: JSON.stringify({
                session_id: session_id,
                message: message,
                functions: functions
            })
        });

        if (response.ok) {
            const data = await response.json();
            if (data.success) {
                return data;
            }
        }

        throw new Error(`Failed to send message: ${response.statusText}`);
    }

    async updateSessionSettings(session_id, settings) {
        const response = await fetch(`${this.base_url}/update-session/`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'X-CSRFToken': this.csrf_token
            },
            body: JSON.stringify({
                session_id: session_id,
                model: settings.model,
                temperature: settings.temperature,
                top_p: settings.top_p,
                max_tokens: settings.max_tokens,
                system_prompt: settings.system_prompt,
                web_search: settings.web_search
            })
        });

        if (response.ok) {
            const data = await response.json();
            return data.success || false;
        }

        return false;
    }
}

class ChatManager {
    constructor(csrf_token) {
        this.api = new ChatAPI(csrf_token);
        this.sessions = new Map();
    }

    createChatSession(chat_id, settings) {
        const session = new ChatSession("", settings, []);
        this.sessions.set(chat_id, session);
        return session;
    }

    async initializeSession(chat_id) {
        if (!this.sessions.has(chat_id)) {
            throw new Error(`Chat session ${chat_id} not found`);
        }

        const session = this.sessions.get(chat_id);
        const session_id = await this.api.createSession(session.settings);
        session.session_id = session_id;

        console.log(`Session initialized for chat ${chat_id}: ${session_id}`);
        return session_id;
    }

    async sendMessage(chat_id, message) {
        if (!this.sessions.has(chat_id)) {
            throw new Error(`Chat session ${chat_id} not found`);
        }

        const session = this.sessions.get(chat_id);

        // Если нет session_id, инициализируем сессию
        if (!session.session_id) {
            await this.initializeSession(chat_id);
        }

        // Добавляем сообщение пользователя
        const user_message = new ChatMessage(
            Date.now().toString(),
            'user',
            message,
            new Date().toISOString()
        );
        session.messages.push(user_message);

        // Отправляем сообщение через API
        const response = await this.api.sendMessage(
            session.session_id,
            message,
            session.settings.functions
        );

        // Добавляем ответ ассистента
        const assistant_message = new ChatMessage(
            (Date.now() + 1).toString(),
            'assistant',
            response.assistant_message.content,
            response.assistant_message.timestamp,
            response.assistant_message.token_stats
        );
        session.messages.push(assistant_message);

        // Обновляем статистику
        if (response.session_stats) {
            session.stats = response.session_stats;
        }

        console.log(`Message sent to chat ${chat_id}, response length: ${assistant_message.content.length}`);
        console.log(`Session settings for ${chat_id}:`, {
            model: session.settings.model,
            web_search: session.settings.web_search
        });
        return assistant_message;
    }

    async updateSettings(chat_id, settings) {
        if (!this.sessions.has(chat_id)) {
            throw new Error(`Chat session ${chat_id} not found`);
        }

        const session = this.sessions.get(chat_id);
        
        // Обновляем локальные настройки
        session.settings = settings;

        // Если есть session_id, обновляем настройки на сервере
        if (session.session_id) {
            const success = await this.api.updateSessionSettings(session.session_id, settings);
            if (success) {
                console.log(`Settings updated for chat ${chat_id}:`, {
                    model: settings.model,
                    web_search: settings.web_search
                });
            }
            return success;
        }

        console.log(`Local settings updated for chat ${chat_id}:`, {
            model: settings.model,
            web_search: settings.web_search
        });
        return true;
    }

    getSession(chat_id) {
        return this.sessions.get(chat_id);
    }

    clearMessages(chat_id) {
        if (this.sessions.has(chat_id)) {
            this.sessions.get(chat_id).messages = [];
            console.log(`Messages cleared for chat ${chat_id}`);
        }
    }

    hasSession(chat_id) {
        return this.sessions.has(chat_id);
    }

    getSessionId(chat_id) {
        const session = this.sessions.get(chat_id);
        return session ? session.session_id : null;
    }

    async forceUpdateSessionSettings(chat_id) {
        if (!this.sessions.has(chat_id)) {
            throw new Error(`Chat session ${chat_id} not found`);
        }

        const session = this.sessions.get(chat_id);
        
        if (session.session_id) {
            const success = await this.api.updateSessionSettings(session.session_id, session.settings);
            if (success) {
                console.log(`Force updated settings for chat ${chat_id}:`, {
                    model: session.settings.model,
                    web_search: session.settings.web_search,
                    system_prompt: session.settings.system_prompt ? 'present' : 'empty',
                    functions: session.settings.functions.length
                });
            }
            return success;
        }
        
        return false;
    }

    async updateAllSessions() {
        const results = {};
        
        for (const [chat_id, session] of this.sessions) {
            if (session.session_id) {
                try {
                    const success = await this.api.updateSessionSettings(session.session_id, session.settings);
                    results[chat_id] = success;
                    if (success) {
                        console.log(`Updated session ${chat_id}:`, {
                            model: session.settings.model,
                            web_search: session.settings.web_search,
                            system_prompt: session.settings.system_prompt ? 'present' : 'empty',
                            functions: session.settings.functions.length
                        });
                    }
                } catch (error) {
                    console.error(`Error updating session ${chat_id}:`, error);
                    results[chat_id] = false;
                }
            } else {
                results[chat_id] = true; // No session to update
            }
        }
        
        return results;
    }
}

class ChatFactory {
    static createChatManager(csrf_token) {
        return new ChatManager(csrf_token);
    }

    static createSettings(model, temperature, top_p, max_tokens, system_prompt = "", web_search = false, functions = []) {
        return new ChatSettings(model, temperature, top_p, max_tokens, system_prompt, web_search, functions);
    }
}

// Экспортируем классы для использования в других модулях
window.ChatManager = ChatManager;
window.ChatFactory = ChatFactory;
window.ChatSettings = ChatSettings;
window.ChatMessage = ChatMessage;
window.ChatSession = ChatSession;
